File: src/main/scala/codechicken/microblock/api/MicroMaterial.scala
package codechicken.microblock.api

import codechicken.lib.render.CCRenderState
import codechicken.lib.render.pipeline.IVertexOperation
import codechicken.lib.vec.{Cuboid6, Matrix4}
import codechicken.microblock.{MicroBlockGenerator, MicroblockClient}
import com.mojang.blaze3d.matrix.MatrixStack
import net.minecraft.block.SoundType
import net.minecraft.client.renderer.model.ItemCameraTransforms.TransformType
import net.minecraft.client.renderer.texture.TextureAtlasSprite
import net.minecraft.client.renderer.{IRenderTypeBuffer, RenderType}
import net.minecraft.entity.Entity
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.item.ItemStack
import net.minecraft.util.math.BlockPos
import net.minecraft.util.text.ITextComponent
import net.minecraft.world.{Explosion, IWorldReader}
import net.minecraftforge.api.distmarker.{Dist, OnlyIn}
import net.minecraftforge.event.RegistryEvent
import net.minecraftforge.registries.ForgeRegistryEntry

/**
 * A MicroMaterial! Used to define the type of a Microblock, can control various aspects
 * of how the MicroMaterial behaves. MixinTraits are supported, Please see [[MicroBlockGenerator]]
 * if you require advanced functionality on the base [[codechicken.multipart.api.part.TMultiPart]] instance.
 *
 * To Register a [[MicroMaterial]] simply subscribe to the [[RegistryEvent.Register]] event with the generic if [[MicroMaterial]],
 * A default implementation of [[MicroMaterial]], [[BlockMicroMaterial]] exits for your convenience.
 * RegistryNames for [[MicroMaterial]] instances should follow the format laid out in [[BlockMicroMaterial.makeMaterialKey()]],
 * registry overrides are supported by [[MicroMaterial]], so you can feel free to add extra functionality at will.
 *
 *
 * TODO, Many of these methods need extra context for forge overloaded methods.
 * Interface for defining a micro material
 */
abstract class MicroMaterial extends ForgeRegistryEntry[MicroMaterial] {

    /**
     * The icon to be used for breaking particles on side
     */
    @OnlyIn(Dist.CLIENT)
    def getBreakingIcon(side: Int): TextureAtlasSprite

    /**
     * Callback to load icons from the underlying block/etc
     */
    @OnlyIn(Dist.CLIENT)
    def loadIcons() {}

    /**
     * This function must return a list of vertex operations, one set for each
     * rendering pass on the preloaded model.
     *
     * @param side   The side that is being rendered as EnumFacing indexes
     * @param layer  The current render layer, null for inventory rendering
     * @param bounds The cuboid bounds of the face being rendered
     */
    @OnlyIn(Dist.CLIENT)
    def getMicroRenderOps(side: Int, layer: RenderType, bounds: Cuboid6): Seq[Seq[IVertexOperation]]

    /**
     * Called to override the item renderer for a given Microblock for your MicroMaterial.
     *
     * @param ccrs          The CCRenderState, already setup, lightmap and overlay can be retrieved from here.
     * @param stack         The ItemStack being rendered.
     * @param transformType The transformation Type of where we are rendering.
     * @param mStack        The Original MatrixStack passed to the ItemRenderer.
     * @param buffers       The IRenderTypeBuffer to get your buffers from.
     * @param mat           A Matrix4 with mStack already applied, and translated to the part center.
     * @param part          The Microblock to render.
     * @return Returning true here means you did something custom and no additional rendering will be done for your item. False means default rendering will be performed.
     *
     */
    @OnlyIn(Dist.CLIENT)
    def renderItem(ccrs: CCRenderState, stack: ItemStack, transformType: TransformType, mStack: MatrixStack, buffers: IRenderTypeBuffer, mat: Matrix4, part: MicroblockClient): Boolean = false

    /**
     * Get the render pass for which this material renders in.
     */
    def canRenderInLayer(layer: RenderType) = layer == RenderType.solid


    /**
     * Gets the render pass for which this material renders as an item.
     */
    def getRenderLayer: RenderType = RenderType.solid

    /**
     * Return true if this material is not opaque (glass, ice).
     */
    def isTransparent: Boolean

    /**
     * Return the light level emitted by this material (glowstone)
     */
    def getLightValue: Int

    /**
     * Return the strength of this material
     */
    def getStrength(player: PlayerEntity): Float

    /**
     * Return the localised name of this material (normally the block name)
     */
    def getLocalizedName: ITextComponent

    /**
     * Get the item that this material is cut from (full block -> slabs)
     */
    def getItem: ItemStack

    /**
     * Get the strength of saw requried to cut this material
     */
    def getCutterStrength: Int

    /**
     * Get the breaking/walking sound
     */
    def getSound: SoundType

    /**
     * Get the explosion resistance of this part to an explosion caused by entity
     */
    def explosionResistance(world: IWorldReader, pos: BlockPos, explosion: Explosion): Float
}

File: src/main/java/codechicken/microblock/api/MicroMaterial.java
package codechicken.microblock.api;

import codechicken.microblock.CBMicroblock;
import codechicken.microblock.util.MicroMaterialRegistry;
import com.mojang.serialization.Codec;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Holder;
import net.minecraft.core.Registry;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.chat.Component;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.resources.ResourceKey;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.BlockGetter;
import net.minecraft.world.level.Explosion;
import net.minecraft.world.level.block.SoundType;
import net.neoforged.fml.loading.FMLEnvironment;
import net.neoforged.neoforge.data.loading.DatagenModLoader;
import net.neoforged.neoforge.registries.RegistryBuilder;
import org.jetbrains.annotations.Nullable;

import java.util.Objects;
import java.util.function.Consumer;

/**
 * Created by covers1624 on 26/6/22.
 */
public abstract class MicroMaterial {

    /**
     * The registry name used by MicroMaterial.
     */
    public static final ResourceKey<Registry<MicroMaterial>> MULTIPART_TYPES = ResourceKey.createRegistryKey(
            ResourceLocation.fromNamespaceAndPath(CBMicroblock.MOD_ID, "micro_material")
    );
    public static final Registry<MicroMaterial> REGISTRY = new RegistryBuilder<>(MULTIPART_TYPES)
            .sync(true)
            .create();

    public static final Codec<MicroMaterial> CODEC = REGISTRY.byNameCodec();
    public static final StreamCodec<RegistryFriendlyByteBuf, MicroMaterial> STREAM_CODEC = ByteBufCodecs.registry(MULTIPART_TYPES);

    @Nullable
    Object renderProperties;

    public MicroMaterial() {
        initClient();
    }

    /**
     * @return Key this material is registered under
     */
    public ResourceLocation getRegistryName() {
        return Objects.requireNonNull(MicroMaterialRegistry.microMaterials().getKey(this));
    }

    /**
     * @return If this material is not opaque. (Glass, Ice, etc.)
     */
    public abstract boolean isTransparent();

    /**
     * Gets the light level emitted by this micro material.
     *
     * @return The light emission.
     */
    public abstract int getLightEmission();

    /**
     * Gets the Strength of this material when being broken
     * by the given player.
     *
     * @param player The player.
     * @return The strength value.
     */
    public abstract float getStrength(Player player);

    /**
     * Gets the localized name for this material.
     *
     * @return The localized name.
     */
    public abstract Component getLocalizedName();

    /**
     * Gets {@link ItemStack} this material can be cut from.
     *
     * @return The {@link ItemStack}.
     */
    public abstract ItemStack getItem();

    /**
     * Get the breaking/waking sound.
     *
     * @return The {@link SoundType}.
     */
    @Nullable
    public abstract SoundType getSound();

    /**
     * Get the resistance of this material for the given explosion.
     *
     * @param level     The level.
     * @param pos       The position.
     * @param explosion The explosion.
     * @return The resistance.
     */
    public abstract float getExplosionResistance(BlockGetter level, BlockPos pos, Explosion explosion);

    /**
     * Initialize any client-side properties for this MicroMaterial.
     *
     * @param cons Consumer to set this material's {@link MicroMaterialClient}.
     */
    public void initializeClient(Consumer<MicroMaterialClient> cons) {
    }

    private void initClient() {
        if (FMLEnvironment.dist.isClient() && !DatagenModLoader.isRunningDataGen()) {
            initializeClient(props -> renderProperties = props);
        }
    }
}
